File: backend/services/catalog/app/schemas/manager.py
from datetime import date
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

from app.schemas.common import PaginatedResponse

CancellationPolicyTypeStr = Literal["FULL", "PARTIAL", "NON_REFUNDABLE"]


# ── Manager Hotels list ───────────────────────────────────


class ManagerHotelItem(BaseModel):
    """Matches the Hotel interface expected by frontend/travel-hub/app/manager/hotels/page.tsx"""

    id: UUID
    name: str
    location: str
    totalRooms: int
    occupiedRooms: int
    status: Literal["ACTIVE", "PENDING_REVIEW"]
    imageUrl: str | None
    categories: int


ManagerHotelListOut = PaginatedResponse[ManagerHotelItem]


# ── Hotel metrics ─────────────────────────────────────────


class HotelStatsOut(BaseModel):
    """Matches the HotelStats interface expected by HotelDetailView.tsx"""

    occupancyRate: float
    activeBookings: int
    monthlyRevenue: float


# ── Room types with availability ──────────────────────────


class RoomTypeManagerItem(BaseModel):
    """Matches the RoomTypeItem interface expected by HotelDetailView.tsx"""

    id: UUID
    name: str
    icon: str
    available: int
    total: int
    rate_plan_id: UUID | None = None


RoomTypeManagerListOut = PaginatedResponse[RoomTypeManagerItem]


# ── Room type promotion ───────────────────────────────────


class RoomTypePromotionOut(BaseModel):
    id: UUID
    rate_plan_id: UUID
    name: str
    discount_type: str
    discount_value: Decimal
    start_date: date
    end_date: date
    is_active: bool


# ── Create promotion ──────────────────────────────────────


class CreatePromotionIn(BaseModel):
    rate_plan_id: UUID
    name: str
    discount_type: Literal["PERCENT", "FIXED"]
    discount_value: Decimal
    start_date: date
    end_date: date


class PromotionCreatedOut(BaseModel):
    id: UUID
    name: str
    discount_type: str
    discount_value: Decimal
    start_date: date
    end_date: date
    is_active: bool


# ── Cancellation policy ───────────────────────────────────


class RatePlanCancellationPolicyOut(BaseModel):
    """Current cancellation policy of a rate plan."""

    type: CancellationPolicyTypeStr
    refund_percent: int | None = None
    hours_limit: int | None = None


class UpdateCancellationPolicyIn(BaseModel):
    """Body for PATCH /manager/rate-plans/{id}/cancellation-policy."""

    type: CancellationPolicyTypeStr
    refund_percent: int | None = None

File: frontend/travel-hub/app/lib/types/manager.ts
import type { RoomTypeIcon } from '@/app/manager/hotels/_data';

export interface ManagerHotelItem {
  id: string;
  name: string;
  location: string;
  totalRooms: number;
  occupiedRooms: number;
  status: 'ACTIVE' | 'PENDING_REVIEW';
  imageUrl: string | null;
  categories: number;
  hotelId: string;
}

export interface HotelStatsOut {
  occupancyRate: number;
  activeBookings: number;
  monthlyRevenue: number;
}

export interface RoomTypeManagerItem {
  id: string;
  name: string;
  icon: RoomTypeIcon;
  available: number;
  total: number;
  rate_plan_id: string | null;
}

export interface RoomTypePromotionOut {
  id: string;
  rate_plan_id: string;
  name: string;
  discount_type: 'PERCENT' | 'FIXED';
  discount_value: number;
  start_date: string;
  end_date: string;
  is_active: boolean;
}

export interface CreatePromotionPayload {
  rate_plan_id: string;
  name: string;
  discount_type: 'NONE' | 'PERCENT' | 'FIXED';
  discount_value: number;
  start_date: string;
  end_date: string;
}

export interface PromotionCreatedOut {
  id: string;
  name: string;
  discount_type: string;
  discount_value: number;
  start_date: string;
  end_date: string;
  is_active: boolean;
}

export type CancellationPolicyTypeStr = 'FULL' | 'PARTIAL' | 'NON_REFUNDABLE';

export interface RatePlanCancellationPolicy {
  type: CancellationPolicyTypeStr;
  refund_percent: number | null;
  hours_limit: number | null;
}

export interface UpdateCancellationPolicyPayload {
  type: CancellationPolicyTypeStr;
  refund_percent?: number;
}

export type PolicyCategory = 'CHECK_IN' | 'CHECK_OUT' | 'PETS' | 'SMOKING' | 'CHILDREN' | 'GENERAL';

export interface PropertyPolicyItem {
  category: PolicyCategory;
  description: string;
}

export type ManagerPropertyImage = {
  id: string;
  url: string;
  caption: string | null;
  display_order: number;
};

export type HotelProfile = {
  id: string;
  name: string;
  description: string | null;
  city: string;
  country: string;
  amenity_codes: string[];
  policies: PropertyPolicyItem[];
  images: ManagerPropertyImage[];
};

export interface TariffBase {
  room_type_id: string;
  base_price: number;
  weekend_premium: number;
}

export interface TariffSeasonalRule {
  id: string;
  room_type_id: string;
  name: string;
  start_date: string;
  end_date: string;
  adjustment_type: 'PERCENT' | 'FIXED';
  adjustment_value: number;
}

export interface RoomTariffs {
  base: TariffBase | null;
  seasonal_rules: TariffSeasonalRule[];
}

export interface UpdateBaseTariffPayload {
  base_price: number;
  weekend_premium: number;
}

export interface AddSeasonalTariffPayload {
  name: string;
  start_date: string;
  end_date: string;
  adjustment_type: 'PERCENT' | 'FIXED';
  adjustment_value: number;
}
